src/diffusers/pipelines/pipeline_utils.py

```diff
@@ -137,6 +146,46 @@ def is_safetensors_compatible(info) -> bool:
     return is_safetensors_compatible
 
 
+def variant_compatible_siblings(info, variant=None) -> Union[List[os.PathLike], str]:
+    filenames = set(sibling.rfilename for sibling in info.siblings)
+    weight_names = [WEIGHTS_NAME, SAFETENSORS_WEIGHTS_NAME, FLAX_WEIGHTS_NAME, ONNX_WEIGHTS_NAME]
+
+    if is_transformers_available():
+        weight_names += [TRANSFORMERS_WEIGHTS_NAME, TRANSFORMERS_SAFE_WEIGHTS_NAME, TRANSFORMERS_FLAX_WEIGHTS_NAME]
+
+    # model_pytorch, diffusion_model_pytorch, ...
+    weight_prefixes = [w.split(".")[0] for w in weight_names]
+    # .bin, .safetensors, ...
+    weight_suffixs = [w.split(".")[-1] for w in weight_names]
+
+    variant_file_regex = (
+        re.compile(f"({'|'.join(weight_prefixes)})({variant})({'|'.join(weight_suffixs)})")
+        if variant is not None
+        else None
+    )
+    non_variant_file_regex = re.compile(f"{'|'.join(weight_names)}")
+
+    if variant is not None:
+        variant_filenames = set(f for f in filenames if variant_file_regex.match(f.split("/")[-1]) is not None)
+    else:
+        variant_filenames = set()
+
+    non_variant_filenames = set(f for f in filenames if non_variant_file_regex.match(f.split("/")[-1]) is not None)
+
+    usable_filenames = set(variant_filenames)
+    for f in non_variant_filenames:
+        variant_filename = f"{f.split('.')[0]}.{variant}.{f.split('.')[1]}"
+        if variant_filename not in usable_filenames:
+            usable_filenames.add(f)
+
+    if len(variant_filenames) > 0 and usable_filenames != variant_filenames:
+        logger.warn(
+            f"\nA mixture of {variant} and non-{variant} filenames will be loaded.\nLoaded {variant} filenames:\n[{', '.join(variant_filenames)}]\nLoaded non-{variant} filenames:\n[{', '.join(usable_filenames - variant_filenames)} from repository files: {', '.join(filenames)}]\nIf this behavior is not expected, please check your folder structure."
```

Comment:
I'm not sure I understand this logic, so I created [this example](https://huggingface.co/pcuenq/v14-variants/tree/main) to test. It contains safetensors and pickle files for both the unnamed `float32` variant and the `float16` one, **but** diffusion_pytorch_model.float16.safetensors is missing from the `unet` folder (we have the 32-bit pickle and safetensors versions, but only the pickle version of the float16 weights): https://huggingface.co/pcuenq/v14-variants/tree/main/unet

In this situation, if we attempt to load the pipeline:

```Python
pipe = DiffusionPipeline.from_pretrained(model_id, variant="float16", torch_dtype=torch.float16)
```

I get the following warning:

```
A mixture of float16 and non-float16 filenames will be loaded.
Loaded float16 filenames:
[text_encoder/model.float16.safetensors, safety_checker/model.float16.safetensors, unet/diffusion_pytorch_model.float16.bin, vae/diffusion_pytorch_model.float16.safetensors, text_encoder/pytorch_model.float16.bin, safety_checker/pytorch_model.float16.bin, vae/diffusion_pytorch_model.float16.bin]
Loaded non-float16 filenames:
[unet/diffusion_pytorch_model.safetensors from repository files: vae/diffusion_pytorch_model.safetensors, text_encoder/model.float16.safetensors, vae/diffusion_pytorch_model.bin, feature_extractor/preprocessor_config.json, model_index.json, vae/diffusion_pytorch_model.float16.safetensors, vae/config.json, unet/diffusion_pytorch_model.safetensors, safety_checker/pytorch_model.float16.bin, safety_checker/pytorch_model.bin, safety_checker/model.float16.safetensors, safety_checker/config.json, vae/diffusion_pytorch_model.float16.bin, text_encoder/model.safetensors, text_encoder/pytorch_model.bin, tokenizer/vocab.json, unet/diffusion_pytorch_model.float16.bin, tokenizer/merges.txt, unet/diffusion_pytorch_model.bin, unet/config.json, tokenizer/special_tokens_map.json, text_encoder/config.json, scheduler/scheduler_config.json, text_encoder/pytorch_model.float16.bin, .gitattributes, safety_checker/model.safetensors, tokenizer/tokenizer_config.json]
If this behavior is not expected, please check your folder structure.
unet/diffusion_pytorch_model.float16.safetensors not found
```

The final comment is totally accurate (there's no `diffusion_pytorch_model.float16.safetensors`), but the rest is confusing. It sounds like the code is going to load the `32-bit` `safetensors` file, but checking the snapshot's filesystem I can verify that actually the PyTorch `float16` version was downloaded and used. In fact, none of the safetensors versions were used.

So, two questions here:
- Should we use the `safetensors` serializations for files that have them, even if some are available as `safetensors` and others aren't? (I understand this might be complicated).
- If we default to everything PyTorch, as we are doing now, what does the previous warning convey?

Reply:
That's a nice edge case! Thanks for checking - I'll correct the warning :-) 

What happens here is that the first warning is not yet aware that .safetensors is not compatible with .bin and therefore we can't load the safetensors. The warning should in general only display either `.bin` or `.safetensors` => will correct! Thanks

Reply:
Should be corrected now, can you test again?